app/Platform/Actions/ResumePlayerSessionAction.php

```diff
@@ -29,34 +29,52 @@ public function execute(
         ?string $userAgent = null,
         ?string $ipAddress = null,
     ): PlayerSession {
+        $sessionKeepAliveTimeInMinutes = 10;
+        $isBackdated = ($timestamp && $timestamp->diffInMinutes(Carbon::now(), true) > $sessionKeepAliveTimeInMinutes);
+
         $timestamp ??= Carbon::now();
 
+        /** @var ?PlayerSession $playerSession */
+        $playerSession = null;
+
         // upsert player game and update last played date right away
         $playerGame = app()->make(AttachPlayerGameAction::class)
             ->execute($user, $game);
-        $playerGame->last_played_at = $timestamp;
-        $playerGame->save();
+        if (!$playerGame->last_played_at || $timestamp > $playerGame->last_played_at) {
+            $playerGame->last_played_at = $timestamp;
+            $playerGame->save();
+        }
 
         $isMultiDiscGameHash = $gameHash?->isMultiDiscGameHash();
 
-        // look for an active session
-        /** @var ?PlayerSession $playerSession */
-        $playerSession = $user->playerSessions()
-            ->where('game_id', $game->id)
-            ->where(function ($query) use ($gameHash, $isMultiDiscGameHash) {
-                if ($gameHash && !$isMultiDiscGameHash) {
-                    $query->where('game_hash_id', $gameHash->id)
-                        ->orWhereNull('game_hash_id');
-                }
-            })
-            ->where(function ($query) use ($userAgent) {
-                if ($userAgent) {
-                    $query->where('user_agent', $userAgent)
-                        ->orWhereNull('user_agent');
-                }
-            })
-            ->orderByDesc('id')
-            ->first();
+        // if a historical session was found, adjust the duration so it includes the timestamp and return it
```

Comment:
nit: recommend changing "if a historical session was found" to "if a historical session **_is_** found". Unless I'm misunderstanding, when we arrive to this comment, we won't have found anything yet - we're about to search for it.